test/models/User.spec.coffee

```diff
@@ -0,0 +1,54 @@
+User  = require 'app/models/User.coffee'
+
+require '../db.coffee'
+
+profile = null
+
+checkUserExists = (id) ->
+	User
+		.findOne id: id
+		.exec()
+		.then (found) ->
+			unless found
+				throw new Error 'User not found'
+
+describe 'User', ->
+	describe '#fromOAuthProfile', ->
+
+		beforeEach ->
+			profile =
+				id: 'id'
+				provider: 'provider'
+				username: 'username'
+
+		afterEach (done) ->
+			User.remove {}, -> done()
+
+		it "should create new user if user with the same id doesn't exist", ->
+			yield User.fromOAuthProfile profile
+			yield checkUserExists 'providerid'
+
+		it 'should return existing user', ->
+			newUser      = yield User.fromOAuthProfile profile
+			existingUser = yield User.fromOAuthProfile profile
+
+			newUser._id.should.deep.equal existingUser._id
+
+		it 'should join provider name with profile id to create userid', ->
+			user = yield User.fromOAuthProfile profile
+			user.id.should.equal 'providerid'
+
+		it 'should assign profile username to user displayName', ->
+			user = yield User.fromOAuthProfile profile
+			user.displayName.should.equal profile.username
+
+		it 'should assign profile username to user displayName even if profile has displayName', ->
+			profile.displayName = 'displayName'
+			user = yield User.fromOAuthProfile profile
+			user.displayName.should.equal profile.username
```

Comment:
Well, `profile.username` is not guaranteed to be different from `'displayName'`
But that's trifle anyway